src/io/flutter/server/vmService/VMServiceManager.java

```diff
@@ -121,14 +120,32 @@ public void onError(RPCError error) {
     });
   }
 
-  public void addRegisteredExtensionRPCs(Isolate isolate) {
-    ApplicationManager.getApplication().invokeLater(() -> {
-      if (isolate.getExtensionRPCs() != null) {
-        for (String extension : isolate.getExtensionRPCs()) {
-          addServiceExtension(extension);
+  public void addRegisteredExtensionRPCs(Isolate isolate, boolean attach) {
+    // If attach was called, there is a risk we may never receive a
+    // Flutter.Frame or Flutter.FirstFrame event so we need to query the
+    // framework to determine if a frame has already been rendered.
+    // This check would be safe to do outside of attach mode but is not needed.
+    if (attach && isolate.getExtensionRPCs() != null && !firstFrameEventReceived) {
+      final EvalOnDartLibrary flutterLibrary = new EvalOnDartLibrary(
+        "package:flutter/src/widgets/binding.dart",
+        vmService,
+        this
+      );
+      flutterLibrary.eval("WidgetsBinding.instance.debugDidSendFirstFrameEvent", null, null).whenCompleteAsync((v, e) -> {
+        // If there is an error we assume the first frame has been received.
+        boolean didSendFirstFrameEvent = e == null ||
+                                         v == null ||
+                                         Objects.equals(v.getValueAsString(), "true");
+        if (didSendFirstFrameEvent) {
+          onFrameEventReceived();
         }
+      });
+    }
+    if (isolate.getExtensionRPCs() != null) {
```

Comment:
Should this be else if?

Reply:
no.  I should add an extra line break to make it clearer.